cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -745,15 +755,15 @@ public void checkCustomCaSecret(CertificateAuthority ca, Secret certSecret, Secr
          * Set inter.broker.protocol.version if it's not set
          * Perform a rolling update.
          */
-        private Future<StatefulSet> kafkaUpgradePhase1(StatefulSet sts, KafkaUpgrade upgrade, String upgradedImage) {
+        private Future<StatefulSet> kafkaUpgradePhase1(StatefulSet sts, ConfigMap cm, KafkaUpgrade upgrade, String upgradedImage) {
             log.info("{}: {}, phase 1", reconciliation, upgrade);
 
             Map<String, String> annotations = Annotations.annotations(sts);
-            Map<String, String> env = ModelUtils.getKafkaContainerEnv(sts);
-            String string = env.getOrDefault(ENV_VAR_KAFKA_CONFIGURATION, "");
-            log.debug("Current config {}", string);
-            KafkaConfiguration currentKafkaConfig = KafkaConfiguration.unvalidated(string);
-            String oldMessageFormat = currentKafkaConfig.getConfigOption(LOG_MESSAGE_FORMAT_VERSION);
+            String config = cm.getData().getOrDefault("server.config", "");
+
+            String oldMessageFormatConfiguration = Arrays.stream(config.split(System.lineSeparator())).filter(line -> line.startsWith(LOG_MESSAGE_FORMAT_VERSION + "=")).findFirst().orElse(null);
+            String oldMessageFormat = oldMessageFormatConfiguration == null ? null : oldMessageFormatConfiguration.substring(oldMessageFormatConfiguration.lastIndexOf("=") + 1);
+
```

Comment:
I think this is a valid point. But I do not think this does anything differently than the previous version did. The old version took the configuration from the STS. So it assumed that the pods were properly rolled after the last update. Only if it would have been read from each pod, it would be the actual value under all circumstances. This version takes the same assumption, just a different source. Because both the STS and config map change trigger the rolling update nd need it to be finished for it to be applied.

So I would suggest to keep this as it is and try to address it better in one of the next steps when we actually talk with the broker about its configuration.

Reply:
> So it assumed that the pods were properly rolled after the last update. 

That what `waitForQuiescence()` was for: Basically block until we know the pods match the STS.

But I'm OK with leaving this until the next PR. It would be better to check the pods themselves, if we could make that work.

Reply:
Ahh, ok. (CC @stanlyDoge since you are looking into the next steps to this)